File: Datastructures/queue/queue.go
/*
Implementation of queue in go language using slices.

`go run queue.go`
*/

package main

import (
        "fmt"
       )


type queue struct{
    list []int
}


func (q queue) is_empty() bool{
    return len(q.list) == 0
}


func (q queue) size() int{
    return len(q.list)
}


func (q *queue) enqueue(value int){
    q.list = append(q.list, value)
}


func (q *queue) dequeue() []int{
    q.list = q.list[1:]
    return q.list
}


func (q queue) print_queue() {
    fmt.Println(q.list)
}


func main() {
    q := queue{}

    v := fmt.Println 
    v(q)
    v(q.size())
    v(q.is_empty())
    q.enqueue(3)
    q.enqueue(4)
    q.enqueue(5)
    q.print_queue()
    q.dequeue()
    q.print_queue()
}

File: Datastructures/queue/queue.swift
/* Implementation of syack in swift

`swift -sdk $(xcrun --show-sdk-path --sdk macosx) queue.swift`
*/


class queue{

    var list = [Int]()

    func is_empty() -> Bool{
        if list.count == 0{
            return true
        }
        else{
            return false
        }
    }

    func size() -> Int{
        return list.count
    }

    func enqueue(value: Int){
        list.append(value)
    }

    func dequeue() -> Int{
        return list.removeFirst()
    }

    func print_queue(){
        print(list)
    }
}

var s = queue()
print(s.is_empty())
print(s.size())
s.enqueue(2)
s.enqueue(4)
s.print_queue()
s.dequeue()
s.print_queue()
